python/lsst/pipe/base/graph/_versionDeserializers.py

```diff
@@ -0,0 +1,524 @@
+# This file is part of pipe_base.
+#
+# Developed for the LSST Data Management System.
+# This product includes software developed by the LSST Project
+# (http://www.lsst.org).
+# See the COPYRIGHT file at the top-level directory of this distribution
+# for details of code ownership.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+from __future__ import annotations
+
+__all__ = ("DESERIALIZER_MAP",)
+
+from dataclasses import dataclass
+from types import SimpleNamespace
+from typing import Callable, ClassVar, Mapping, DefaultDict, Set, Dict, Tuple, Optional
+
+import json
+import lzma
+import networkx as nx
+import pickle
+import struct
+import uuid
+
+from collections import defaultdict
+from typing import TYPE_CHECKING
+
+from lsst.utils import doImport
+from lsst.pex.config import Config
+from lsst.daf.butler import DimensionUniverse, Quantum, SerializedDimensionRecord, DimensionRecord
+
+from .quantumNode import QuantumNode, SerializedQuantumNode
+from ..pipeline import TaskDef
+from ..pipelineTask import PipelineTask
+from ._implDetails import _DatasetTracker, DatasetTypeName
+
+
+if TYPE_CHECKING:
+    from .graph import QuantumGraph
+
+
+class StructSizeDescriptor:
+    """This is basically a class level property. It exists to report the size
+    (number of bytes) of whatever the formatter string is for a deserializer
+    """
+    def __get__(self, inst, owner) -> int:
+        return struct.calcsize(owner.FMT_STRING)
+
+
+@dataclass
+class DeserializerBase:
+    FMT_STRING: ClassVar[str] = NotImplemented
```

Comment:
I will change to what you ask, but I fundamentally disagree with it in python.
1) I dont think that attributes should shadow properties to get this sort of behavior
2) I don't think that syntax is very clear to downstream authors that they may implement things via an attribute instead of implementing another property, and I don't really think it reads clearly
3) I think it is an unnecessary introduction of a metaclass, which can have other downstream implications in some cases (not likely in this case)
4) I think this is EXACTLY the kind of thing that `__init_subclass__` was introduced to help you do.
5) If this was any other form of validation, you would have not thought twice, why should this be special

Reply:
FWIW, I agree that attributes should not be used to shadow properties, and I don't think an attribute should ever be used to implement an abstract property (and I think MyPy agrees with me).  One should always write a simpler forwarding property to some private attribute instead (unless you've got evidence that the computational cost of accessing the property is a serious bottleneck _and_ this is the only good way to fix it).  It's a shame Python doesn't have a way to say, "subclasses must provide this public attribute", but I'm not super bothered by that because I find I usually regret public attributes anyway for the same reason they're an antipattern in essentially all other programming langauges (at least on classes that conform to a interface, as opposed to dumb structs).

But anyhow, there is no shadowing or properties in play here, because this is all class-level attribute stuff.

To the real issue: I think I'd just reverse your last statement: if this was any other form of "does this class implement an interface" validation, I think you'd have used `abc` without thinking twice; why should this be special?

Reply:
This may be a pylance difference, I'm going to look into that, but the following seems to have no problem, and ALSO completely satisfies ABC and AFAICT is often used to implement abstract class attributes.
```
class Base(ABC):
    @property
    @abstractmethod
    def FMT_STRING(self) -> str:
        pass

class Impl(Base):
    FMT_STRING: ClassVar[str] = '>Q'
```
I think I disagree with you on the value of properties here. If one does not need them, use an attribute. Properties give the ability to 'upgrade' or refactor your interface later if attributes prove to not do the trick. Starting with one seems to put the cart before the horse needlessly. And from the outside there is no difference at all between a property and an attribute, so I don't find the argument about conforming to a particular interface very compelling, and in particular would hate to go back to getter patterns.

There is shadowing in play here, that is what a subclass is. A definition in a subclass shadows the base class instance by interrupting the MRO (attribute too).

I would be perfectly happy if abc provided a reasonable check for class level attributes.

Reply:
On that code snippet example: fair enough; I consider that a defect, and definitely a broken thing that a reviewer should flag even if `abc` doesn't - but I can believe that `abc` doesn't check it either because it can't [do it efficiently] or because it's not a great implementation.

And I don't advocate starting with properties all the time, but it's rare that I don't find myself switching to them eventually _if the class is part of an inheritance tree_.  I find I need that extra flexibility most of the time, or that having a `property` saves having base class state.

I agree that this is technically shadowing, and at some level the fact that inheritance is a special kind of shadowing is why it has so many rules and hard-won wisdom about how to use it safely.  Shadowing is dangerous; inheritance puts a handle on that sharp knife, but it's still a sharp blade that can do a lot of damage if used carelessly.

> I would be perfectly happy if abc provided a reasonable check for class level attributes.

Agreed.  This debate, many other debates, and many real-world bugs would have been avoided if there was a built-in way to do this.  And while I still maintain that `abc.classmethod` is the least-bad approach (it's just _very_ verbose), I do think custom checking of the kind you did is much better than not checking at all (looking at you, `Task.ConfigClass` and `Task._DefaultName` and who knows what else in our codebase!), especially back in the pre-type-annotation days where you couldn't even declare the attribute without assigning it to something, leaving it super easy for derived classes to forget to override it.

Reply:
The other thing I like about this solution is that it happens at declaration time, rather than at some point later down the line when someone tries to instantiate some abc subclass.

Can we compromise and leave this behavior in (slightly modified), and I will upgrade this to a protocol as well so that this will get type checking errors as well if the attribute is not implemented? I will send you a small sample of what I mean, to see what you think.

Reply:
> The other thing I like about this solution is that it happens at declaration time, rather than at some point later down the line when someone tries to instantiate some abc subclass.

That's the strongest argument in favor so far.  Not enough to convince me that this is a good solution, but it's a big step towards convincing me that an abstract `classmethod` isn't enough of a solution, and we should have a standardized solution in `utils` or something.

> Can we compromise and leave this behavior in (slightly modified), and I will upgrade this to a protocol as well so that this will get type checking errors as well if the attribute is not implemented?

Adding a `Protocol` doesn't really feel like it helps, but that's because I'm pushing back on your solution to this problem so much because
A) it's a very common problem most developers will encounter at one point or another;
B) a solution involving a custom `__init_subclass__` (let alone `Protocol`) is not something I am comfortable recommending to the full team;
C) I am not comfortable with you using one pattern for this problem while recommending a different solution to other members of the team.

Now that I've written that out, though, I've also convinced myself that this needs to be RFC'd and a solution put in the dev guide, because it is a common problem with no standard-pattern solution, and me having this debate in code reviews is not going to converge to a uniform solution to this problem throughout the codebase.  I think I'll start by proposing the abstract `classmethod` approach, but I'll include your point (with credit) that an import-time solution would be better than a construction-time solution, and maybe we'll converge to adding something to `utils` that I would feel comfortable recommending the full team use.

In the meantime, feel free to merge whatever you like; I don't want to hold up the whole ticket on this point.  Just expect it to change as part of the RFC implementation (I'm not going to propose we change all historical solutions to this problem, as that would be hugely disruptive in some cases, but changing this and a few other cases after-the-fact should be fine).

Reply:
I am not sure I agree that I would not recommend `__init_subclass__` for others, but I do agree that some sort of standard is needed, and welcomed. If you think a turn key solution would be better, I would be happy to make a class decorator as a prototype on this ticket, and if we converge on the RFC, we can move it to utils.

Reply:
Once the RFC converges, I'd be happy to let you take the lead on designing something consistent with it, and before that, you're welcome to prototype an alternative to my initial proposal of just using abstract `classmethod` - but I think it probably shouldn't be on this ticket, which we don't want to hold up on a side issue.